### PR TITLE
chore: make retries work with custom timeouts

### PR DESCRIPTION
**Pull request by mxschmitt**
Differences between NUnit and MSTest:

- NUnit will call the `afterEach` when a timeout is reached, this is not possible in MSTest since the `base.Execute()` takes care of calling `beforeEach/afterEach`.

Both implementations will:
- throw if test framework timeout < Playwright test timeout
- gracefully close the BrowserContext when the test timeout has reached
- making reached timeouts compatible with retries

Relates https://github.com/microsoft/playwright-dotnet/issues/2316

### CODE REVIEW

File: src/Playwright.NUnit/TimeoutCommand.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        // cancel by that the underlying BrowserContext calls.
+        if (timeoutTask.IsCompleted)
+        {
+            // 2.1 Call the test teardown handlers.
+            await RunSetUpOrTearDownMethod(context, isTearDown: true);
+            TearDownRan = true;
+
+            // 2.2 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
```
Comment (mxschmitt) [inline src/Playwright.NUnit/TimeoutCommand.cs:52]: look at this

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                }
+            }
+            finally
+            {
+                CleanupMethods.Clear();
+            }
+
+            // 2.2 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
```
Comment (dgozman) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:80]: Drop this one please.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        if (timeout != null && timeout != 0 && playwrightTestTimeout > timeout)
+        {
+            result = new TestResult[]
+            {
+                new TestResult()
+                {
+                    Outcome = UnitTestOutcome.Failed,
+                    TestFailureException = new Exception($"Playwright test timeout ({playwrightTestTimeout}ms) is larger than MSTest timeout ({timeout}ms).")
```
Comment (dgozman) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:114]: Is it smaller or larger? How should it be? I find naming here somewhat confusing.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+internal class PlaywrightTimeoutAttribute : TestMethodAttribute
+{
+    static internal List<Func<Task>> CleanupMethods { get; set; } = new List<Func<Task>>();
+
+    public override TestResult[] Execute(ITestMethod testMethod)
+    {
+        var timeout = PlaywrightSettingsProvider.TestTimeout;
+        if (timeout == null || Debugger.IsAttached)
+        {
+            return base.Execute(testMethod);
```
Comment (dgozman) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:44]: This misses `CleanupMethods.Clear()`.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        await Task.WhenAny(testExecutionTask, timeoutTask);
+
+        // 2. If timeout was reached, call the tearDownHandlers and
+        // cancel by that the underlying BrowserContext calls.
+        if (timeoutTask.IsCompleted)
+        {
+            // 2.1 Call the test teardown handlers.
+            try
+            {
+                foreach (var cleanupMethod in CleanupMethods)
```
Comment (dgozman) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:68]: Shouldn't we call these when timeout did not happen?

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                {
+                    await cleanupMethod();
```
Comment (dgozman) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:70]: So this only cleans up our context, but does not run other `[TestCleanup]` methods. Is there a way to do that?

File: src/Playwright.TestAdapter/MSTestProvider.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+using System.Xml;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+
+namespace Microsoft.Playwright.TestAdapter;
+
+/// <summary>
+/// MSTest does not provide a way to get the global timeout value during a test execution.
```
Comment (dgozman) [inline src/Playwright.TestAdapter/MSTestProvider.cs:32]: Can we get a timeout for the current test instead?

File: src/Playwright.NUnit/TimeoutCommand.cs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            }
+            return;
+        }
+
+        context.CurrentResult = context.CurrentTest.MakeTestResult();
+        context.CurrentResult.RecordException(new TimeoutException(timeoutExceptionMessage));
+        return;
+    }
+
+    protected async Task RunSetUpOrTearDownMethod(TestExecutionContext context, bool isSetUp = false, bool isTearDown = false)
```
Comment (dgozman) [inline src/Playwright.NUnit/TimeoutCommand.cs:74]: I feel uncomfortable with all the internal details of NUnit we are touching here.

File: src/Playwright.NUnit/TimeoutCommand.cs [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        if (PlaywrightTimeoutIsSmallerThanNUnitTimeout(context, timeout.Value, out var result))
+        {
+            return result!;
+        };
+
+        ExecuteWithTimeoutAsync(context, timeout.Value).GetAwaiter().GetResult();
+        return context.CurrentResult;
+    }
+
+    private async Task ExecuteWithTimeoutAsync(TestExecutionContext context, int timeout)
```
Comment (dgozman) [inline src/Playwright.NUnit/TimeoutCommand.cs:34]: This needs a comment that explains whether this function runs setups and teardowns.

File: src/Playwright.NUnit/PlaywrightTestAttribute.cs [hunk 10]
```diff
@@ -76,6 +83,13 @@ public override TestResult Execute(TestExecutionContext context)
                 {
                     break;
                 }
+                // NUnit takes care about running TearDown for the last iteration. PlaywrightTestAttribute inherits from
+                // the TimeoutCommand which runs TearDown for us in a case of a timeout.
+                if (!TearDownRan)
```
Comment (dgozman) [inline src/Playwright.NUnit/PlaywrightTestAttribute.cs:88]: I think merging `RetryTestCommand` and `TimeoutCommand` could lead to easier code, wdyt?

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs (hunk 3 above)
Comment (mxschmitt) [reply src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:114]: I can rephrase it to e.g. `Playwright.TestTimeout is larger than MSTest.TestTimeout, make sure to have a larger Playwright timeout otherwise test execution gets terminated`

File: src/Playwright.NUnit/PlaywrightTestAttribute.cs (hunk 10 above)
Comment (mxschmitt) [reply src/Playwright.NUnit/PlaywrightTestAttribute.cs]: I made an attempt, seems better already.

File: src/Playwright.TestAdapter/MSTestProvider.cs (hunk 7 above)
Comment (mxschmitt) [reply src/Playwright.TestAdapter/MSTestProvider.cs:32]: Its unfortunately not possible: https://cs.github.com/microsoft/testfx/blob/e5bc0e46f6d2cc8f0465596ff77ff18bb6680253/src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs#L718

not exposed to the public, after a discussion with Jakub this seems the best way to determine it.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs (hunk 5 above)
Comment (mxschmitt) [reply src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:75]: we do it already here: https://github.com/mxschmitt/playwright-dotnet/blob/45c4e554b95a422c79afa40b106b5167b83cfb78/src/Playwright.MSTest/ContextTest.cs#L44-L51

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            return result!;
+        }
+
+        return ExecuteWithTimeoutAsync(testMethod, timeout.Value).GetAwaiter().GetResult();
+    }
+
+    private async Task<TestResult[]> ExecuteWithTimeoutAsync(ITestMethod testMethod, int timeout)
+    {
+        // 1. Race Test execution against the specified timeout.
+        var testExecutionTask = Task.Run(() => base.Execute(testMethod));
```
Comment (mxschmitt) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:64]: dig into if we can replace the method to handle the timeout inside so teardown will run after a timeout happened.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            {
+                CleanupMethods.Clear();
+            }
+
+            // 2.2 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
+        }
+
+        if (testExecutionTask.IsCompleted)
```
Comment (mxschmitt) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:90]: swap with below

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    private async Task<TestResult[]> ExecuteWithTimeoutAsync(ITestMethod testMethod, int timeout)
+    {
+        // 1. Race Test execution against the specified timeout.
+        // In MSTest base.Execute will internally run the TestInitialize and TestCleanup. In case of a timeout, we don't
+        // run TestCleanup, only TestInitialize. To get good error messages, we close the browser context inside ContextCloseHookOnTimeout.
+        var testExecutionTask = Task.Run(() => base.Execute(testMethod));
+        var timeoutTask = Task.Delay(timeout);
+        await Task.WhenAny(testExecutionTask, timeoutTask);
```
Comment (nohwnd) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:61]: This returns the task that completed that you don't capture. I am not sure if you don't use it on purpose because it made the code read worse, or if you don't know about it. 

That said I don't see any problem with checking the `.IsCompleted` on the task. The worst thing that I can see happen is that you await the test again in 3. if the timeout and test both completed because ContextCloseHookOnTimeout took enough time for both of them to complete. Wrapping a completed task into another task and awaiting it again has not functional issue, other than a tiny tiny added overhead.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+        // 2. Close the BrowserContext
+        if (ContextCloseHookOnTimeout != null)
+        {
+            await ContextCloseHookOnTimeout().ConfigureAwait(false);
```
Comment (nohwnd) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:66]: Would be nice to see some kind of cancellation of the close hook, but maybe I am overthinking it.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        if (timeout != null && timeout != 0 && playwrightTestTimeout > timeout)
+        {
+            return new TestResult[]
+            {
+                new TestResult()
+                {
+                    Outcome = UnitTestOutcome.Failed,
+                    TestFailureException = new Exception($"Playwright test timeout ({playwrightTestTimeout}ms) is larger than MSTest timeout ({timeout}ms).")
```
Comment (nohwnd) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:109]: Maybe [InvalidOperationException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception) could be used here.

It would be nice to see the more verbose explanation you proposed above to know why Playwright timeout should be larger. I would also use longer rather than larger, as we are referring to time.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+namespace Microsoft.Playwright.MSTest;
+
+internal class PlaywrightTimeoutAttribute : TestMethodAttribute
+{
+    static internal Func<Task>? ContextCloseHookOnTimeout { get; set; } = null;
+
+    public override TestResult[] Execute(ITestMethod testMethod)
+    {
+        var timeout = PlaywrightSettingsProvider.TestTimeout;
+        if (timeout == null || Debugger.IsAttached)
```
Comment (nohwnd) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:42]: I get the intent of not timing out under debugger, but it makes it impossible to debug the timeout flow. It would be nice to have a ENV variable that you can set and make it go into the async flow. You will have to pump up your timeouts in the test if you want to be able to go step through it, but to me it looks useful when you are assisting someone with their issue related to timeouts.

But feel free to ignore my suggestions if you think feature creep is bigger issue than not being able to debug this.

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Playwright.TestAdapter;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Playwright.MSTest;
+
+internal class PlaywrightTimeoutAttribute : TestMethodAttribute
+{
+    static internal Func<Task>? ContextCloseHookOnTimeout { get; set; } = null;
```
Comment (nohwnd) [inline src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:37]: Rename ContextCloseHookOnTimeoutAsync? I can see it being invoked but not awaited being an easy error to make in your code.

File: src/Playwright.NUnit/ContextTest.cs [hunk 18]
```diff
@@ -41,4 +41,11 @@ public async Task ContextSetup()
     {
         Context = await NewContext(ContextOptions()).ConfigureAwait(false);
     }
+
+    [TearDown]
+    public async Task ContextTearDown()
+    {
+        await Context.CloseAsync().ConfigureAwait(false);
+        Context = null!;
```
Comment (nohwnd) [inline src/Playwright.NUnit/ContextTest.cs:49]: I could not find why this is done, and it seems a bit obscure to force break your nullability. Are you checking it for null somewhere, or are you just trying to get it garbage collected earlier?

File: src/Playwright.NUnit/PlaywrightTestAttribute.cs [hunk 19]
```diff
@@ -130,6 +144,117 @@ private static string Indent(string text, int indent)
...
+
+        private (TestResult, bool) ExecuteWithTimeout(TestExecutionContext context)
+        {
+            var timeout = PlaywrightSettingsProvider.TestTimeout;
+            if (timeout == null || Debugger.IsAttached)
+            {
+                return (innerCommand.Execute(context), false);
+            }
+
+            if (CheckThatPlaywrightTimeoutIsSmallerThanNUnitTimeout(context, timeout.Value) is var result && result != null)
```
Comment (nohwnd) [inline src/Playwright.NUnit/PlaywrightTestAttribute.cs:156]: nit: if you don't use `var` the is check implies not null. `is TestResult result` would automatically mean that `result` has value. 

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYAGAWAChqBGRthxgFRgGcMAlXjgA2GAARQxAXjF4RwgNzsKzAJwAKSQEseY7n0E8R42EdEBKJRyZU9vAUNFiA3gF8gA==

Comment (mxschmitt) [reply src/Playwright.NUnit/PlaywrightTestAttribute.cs]: cool

File: src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs (hunk 16 above)
Comment (mxschmitt) [reply src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs:42]: I'll follow-up.
